### ISSUE TITLE
Vanished ships are still valid and accessible (for the frame they were vanished)

### ISSUE THREAD

**Issue by Baezon** [scp-fs2open/fs2open.github.com#5034]
This is mostly an issue for scripts, as sexps appear to mostly use the registry which is properly updated with vanishing.

But once vanished, that ship handle is still valid, the ship still has hitpoints, is still returned by `mn.getShipList()`, and indeed there does not appear to be *any way* to determine that that ship has been vanished. This is, needless to say, highly unintuitive and confusing. 

Given that vanishing doesn't actually do all that terribly much in the moment but simply applies the `Should_be_dead` flag, this result is understandable after some investigation, but I'm not sure if `isValid()` should check for that flag, if the flag should be exposed to scripters, or the script system itself changed to not allow modification or access to such ships at all (that might seem extreme, since destroying the ship in other ways typically still has the ship be accessible because the ship object is truly still present, but given that 'vanish' is the closest thing to an 'instantly delete' button, I don't think its a terribly unreasonable option).

**Comment by BMagnu** [comment]
I would not make "Should_be_dead" the same as invalid.
Neither for the script system, nor for explicit calls to IsValid.
I can imagine that there are scripts which rely on having access to technically dead ships. Pretty much anything in hooks running after ship death, I imagine.
IMO, it should be accessible as a function returning a bool. Currently, checking for the aliveness of a ship involves testing against hitpoints, which is fickle, so having a proper "is dead" check would be good.
That, on the other hand, the ship iterator won't return these ships in the first place is a different story.

TL;DR: Getting a dead ship from a hook or by keeping it in a variable should be possible at least until it gets fully invalidated, but the iterator probably shouldn't return them

**Comment by Baezon** [comment]
This is quite different from the ship being dead, i.e. <= 0 hitpoints, really the flag should be called `Should_be_deleted`. As mentioned, the ship object quite validly persists well after things like $On Death are triggered, who knows how long it could take to complete its death roll. But a `Should_be_dead` ship by contrast, won't even survive the frame before being well and truly invalidated.

This is not about 'dead' ships but rather 'deleted' ships, at least as I see it. As scripter I quite often want to not merely kill but completely delete.

My use case for what it's worth, is essentially like saving, look at the state, store it to file, but it's triggered in this case by an event which also vanishes a ship, and thus when the save happens it erroneously retrieves and stores this ship that shouldn't still exist. Workarounds like manually telling the saving to exempt one or more ships, or to delay the saving by a frame aren't complicated but quite ugly.

**Comment by Goober5000** [comment]
It's possible to access the ship registry within your script and grab the status that way, although that involves a few extra steps.

Keeping a dead ship handle is sometimes needed.  I like the idea of a `shouldBeDead` function on the object type, because this sounds like a problem that could also apply to other things such as weapons.  And I also think it makes sense to remove should-be-dead objects from the various iterators - `mn.getShipList()`, `mn.Ships`, `mn.Weapons`, etc.

**Comment by Baezon** [comment]
I'd definitely motion to rename the flag to `Should_be_deleted` not least of which to disambiguate it from `isDying()` or `hasShipExploded()` if it were exposed.
